File: gateway/app/services/broker.py
import asyncio
import json
import time
from typing import Any, Awaitable, Callable, Dict, Optional

import aio_pika
import redis.asyncio as redis
from aio_pika import ExchangeType
from opentelemetry.trace import SpanKind

from .logging_service import LoggingService
from .message import internal_from_dict
from .metrics import MetricsService
from .settings import Settings
from .tracing import TracingService


class BrokerService:
    def __init__(
        self,
        settings: Settings,
        metrics: MetricsService,
        tracing: TracingService,
        logger: LoggingService,
    ) -> None:
        self._settings = settings
        self._metrics = metrics
        self._tracing = tracing
        self._logger = logger
        self.redis_publish_client: Optional[redis.Redis] = None
        self.rabbit_publish_connection: Optional[aio_pika.RobustConnection] = None
        self.rabbit_publish_channel: Optional[aio_pika.Channel] = None
        self.rabbit_inbox_exchange: Optional[aio_pika.Exchange] = None
        self.rabbit_events_exchange: Optional[aio_pika.Exchange] = None
        self._last_stream_trim: Dict[str, float] = {}
        self._send_to_subjects: Optional[Callable[[list, Any], Awaitable[int]]] = None

    def set_sender(self, sender: Callable[[list, Any], Awaitable[int]]) -> None:
        self._send_to_subjects = sender

    def _strip_internal_payload(self, payload: Any) -> Any:
        if not self._settings.OUTBOX_STRIP_INTERNAL:
            return payload
        if not isinstance(payload, dict):
            return payload
        internal = internal_from_dict(payload)
        if not internal:
            return payload
        return internal.to_client_payload()

    async def startup(self) -> None:
        if self._settings.REDIS_DSN:
            self.redis_publish_client = redis.from_url(self._settings.REDIS_DSN, decode_responses=True)
            if self._settings.ROLE_READ:
                asyncio.create_task(self._redis_outbox_consumer())
        if self._settings.RABBITMQ_DSN:
            asyncio.create_task(self._init_rabbitmq())
            if self._settings.ROLE_READ:
                asyncio.create_task(self._rabbit_outbox_consumer())

    async def shutdown(self) -> None:
        if self.rabbit_publish_connection:
            await self.rabbit_publish_connection.close()
            self.rabbit_publish_connection = None

    def _rabbit_queue_args(self, ttl_ms: int) -> Dict[str, Any]:
        args: Dict[str, Any] = {}
        if ttl_ms > 0:
            args["x-message-ttl"] = ttl_ms
        if self._settings.RABBITMQ_DLQ_EXCHANGE:
            args["x-dead-letter-exchange"] = self._settings.RABBITMQ_DLQ_EXCHANGE
            if self._settings.RABBITMQ_DLQ_QUEUE:
                args["x-dead-letter-routing-key"] = self._settings.RABBITMQ_DLQ_QUEUE
        return args

    async def _push_redis_dlq(self, client: redis.Redis, reason: str, raw: str) -> None:
        try:
            await client.xadd(self._settings.REDIS_DLQ_STREAM, {"reason": reason, "raw": raw})
        except Exception:
            pass

    async def _push_rabbit_dlq(self, channel: aio_pika.Channel, reason: str, raw: bytes) -> None:
        try:
            message = aio_pika.Message(body=raw, headers={"reason": reason})
            dlq_exchange = await channel.declare_exchange(
                self._settings.RABBITMQ_DLQ_EXCHANGE,
                ExchangeType.DIRECT,
                durable=True,
            )
            await dlq_exchange.publish(message, routing_key=self._settings.RABBITMQ_DLQ_QUEUE)
        except Exception:
            pass

    async def _maybe_trim_stream(self, client: redis.Redis, stream: str) -> None:
        if self._settings.REPLAY_STRATEGY != "bounded":
            return
        if self._settings.REPLAY_RETENTION_SECONDS <= 0:
            return
        now = time.time()
        last = self._last_stream_trim.get(stream, 0.0)
        interval = max(1.0, min(60.0, self._settings.REPLAY_RETENTION_SECONDS / 4))
        if now - last < interval:
            return
        min_id = f"{int((now - self._settings.REPLAY_RETENTION_SECONDS) * 1000)}-0"
        try:
            await client.xtrim(stream, minid=min_id)
        except Exception:
            pass
        self._last_stream_trim[stream] = now

    async def publish_broker(
        self,
        stream: Optional[str],
        exchange: Optional[aio_pika.Exchange],
        routing_key: str,
        payload: Dict[str, Any],
        body: Optional[str] = None,
    ) -> None:
        if body is None:
            body = json.dumps(payload, separators=(",", ":"), sort_keys=True)
        if self.redis_publish_client and stream:
            try:
                kwargs: Dict[str, Any] = {}
                if self._settings.REPLAY_STRATEGY == "bounded" and self._settings.REPLAY_MAXLEN > 0:
                    kwargs["maxlen"] = self._settings.REPLAY_MAXLEN
                    kwargs["approximate"] = True
                await self.redis_publish_client.xadd(stream, {"data": body}, **kwargs)
                self._metrics.inc("broker_publish_total")
                await self._maybe_trim_stream(self.redis_publish_client, stream)
            except Exception:
                pass
        if exchange:
            try:
                message = aio_pika.Message(body=body.encode("utf-8"))
                await exchange.publish(message, routing_key=routing_key)
                self._metrics.inc("broker_publish_total")
            except Exception:
                pass

    async def _redis_outbox_consumer(self) -> None:
        if not self._settings.REDIS_DSN:
            return
        client = redis.from_url(self._settings.REDIS_DSN, decode_responses=True)
        last_id = "$" if self._settings.REPLAY_STRATEGY == "none" else "0-0"
        backoff = 1.0
        while True:
            try:
                response = await client.xread({self._settings.REDIS_STREAM: last_id}, block=5000, count=10)
                if not response:
                    backoff = 1.0
                    continue
                for _stream, messages in response:
                    for msg_id, fields in messages:
                        last_id = msg_id
                        raw = fields.get("data", "{}")
                        try:
                            data = json.loads(raw)
                            subjects = data.get("subjects", [])
                            payload = self._strip_internal_payload(data.get("payload"))
                            traceparent = data.get("traceparent") or ""
                            span_ctx = self._tracing.extract_context(traceparent) if traceparent else None
                            if self._tracing.enabled and self._tracing.should_record(bool(traceparent)):
                                with self._tracing.tracer.start_as_current_span(
                                    "ws.outbox.redis", context=span_ctx, kind=SpanKind.CONSUMER
                                ) as span:
                                    span.set_attribute("ws.subjects_count", len(subjects))
                                    if self._send_to_subjects:
                                        await self._send_to_subjects(subjects, payload)
                            else:
                                if self._send_to_subjects:
                                    await self._send_to_subjects(subjects, payload)
                        except Exception:
                            await self._push_redis_dlq(client, "parse_error", raw)
            except Exception:
                await asyncio.sleep(backoff)
                backoff = min(30.0, backoff * 2)

    async def _rabbit_outbox_consumer(self) -> None:
        if not self._settings.RABBITMQ_DSN:
            return
        backoff = 1.0
        while True:
            try:
                connection = await aio_pika.connect_robust(self._settings.RABBITMQ_DSN)
                async with connection:
                    channel = await connection.channel()
                    exchange = await channel.declare_exchange(
                        self._settings.RABBITMQ_EXCHANGE, ExchangeType.DIRECT, durable=True
                    )
                    dlq_exchange = await channel.declare_exchange(
                        self._settings.RABBITMQ_DLQ_EXCHANGE, ExchangeType.DIRECT, durable=True
                    )
                    dlq_queue = await channel.declare_queue(self._settings.RABBITMQ_DLQ_QUEUE, durable=True)
                    await dlq_queue.bind(dlq_exchange, routing_key=self._settings.RABBITMQ_DLQ_QUEUE)
                    queue = await channel.declare_queue(
                        self._settings.RABBITMQ_QUEUE,
                        durable=True,
                        arguments=self._rabbit_queue_args(self._settings.RABBITMQ_QUEUE_TTL_MS),
                    )
                    await queue.bind(exchange, routing_key=self._settings.RABBITMQ_ROUTING_KEY)
                    async with queue.iterator() as queue_iter:
                        async for message in queue_iter:
                            async with message.process():
                                try:
                                    data = json.loads(message.body.decode("utf-8"))
                                    subjects = data.get("subjects", [])
                                    payload = self._strip_internal_payload(data.get("payload"))
                                    traceparent = data.get("traceparent") or ""
                                    span_ctx = self._tracing.extract_context(traceparent) if traceparent else None
                                    if self._tracing.enabled and self._tracing.should_record(bool(traceparent)):
                                        with self._tracing.tracer.start_as_current_span(
                                            "ws.outbox.rabbitmq", context=span_ctx, kind=SpanKind.CONSUMER
                                        ) as span:
                                            span.set_attribute("ws.subjects_count", len(subjects))
                                            if self._send_to_subjects:
                                                await self._send_to_subjects(subjects, payload)
                                    else:
                                        if self._send_to_subjects:
                                            await self._send_to_subjects(subjects, payload)
                                    backoff = 1.0
                                except Exception:
                                    await self._push_rabbit_dlq(channel, "parse_error", message.body)
            except Exception:
                await asyncio.sleep(backoff)
                backoff = min(30.0, backoff * 2)

    async def _init_rabbitmq(self) -> None:
        backoff = 1.0
        while True:
            try:
                self.rabbit_publish_connection = await aio_pika.connect_robust(self._settings.RABBITMQ_DSN)
                self.rabbit_publish_channel = await self.rabbit_publish_connection.channel()
                self.rabbit_inbox_exchange = await self.rabbit_publish_channel.declare_exchange(
                    self._settings.RABBITMQ_INBOX_EXCHANGE,
                    ExchangeType.DIRECT,
                    durable=True,
                )
                self.rabbit_events_exchange = await self.rabbit_publish_channel.declare_exchange(
                    self._settings.RABBITMQ_EVENTS_EXCHANGE,
                    ExchangeType.DIRECT,
                    durable=True,
                )
                if self._settings.RABBITMQ_DLQ_EXCHANGE and self._settings.RABBITMQ_DLQ_QUEUE:
                    dlq_exchange = await self.rabbit_publish_channel.declare_exchange(
                        self._settings.RABBITMQ_DLQ_EXCHANGE,
                        ExchangeType.DIRECT,
                        durable=True,
                    )
                    dlq_queue = await self.rabbit_publish_channel.declare_queue(
                        self._settings.RABBITMQ_DLQ_QUEUE,
                        durable=True,
                    )
                    await dlq_queue.bind(dlq_exchange, routing_key=self._settings.RABBITMQ_DLQ_QUEUE)
                if self._settings.RABBITMQ_INBOX_QUEUE and self.rabbit_inbox_exchange:
                    inbox_queue = await self.rabbit_publish_channel.declare_queue(
                        self._settings.RABBITMQ_INBOX_QUEUE,
                        durable=True,
                        arguments=self._rabbit_queue_args(self._settings.RABBITMQ_INBOX_QUEUE_TTL_MS),
                    )
                    await inbox_queue.bind(self.rabbit_inbox_exchange, routing_key=self._settings.RABBITMQ_INBOX_ROUTING_KEY)
                if self._settings.RABBITMQ_EVENTS_QUEUE and self.rabbit_events_exchange:
                    events_queue = await self.rabbit_publish_channel.declare_queue(
                        self._settings.RABBITMQ_EVENTS_QUEUE,
                        durable=True,
                        arguments=self._rabbit_queue_args(self._settings.RABBITMQ_EVENTS_QUEUE_TTL_MS),
                    )
                    await events_queue.bind(self.rabbit_events_exchange, routing_key=self._settings.RABBITMQ_EVENTS_ROUTING_KEY)
                self._logger.logger.info("rabbitmq.connected")
                return
            except Exception as exc:
                self._logger.logger.warning("rabbitmq.connect_failed: %s", exc)
                await asyncio.sleep(backoff)
                backoff = min(30.0, backoff * 2)

File: rust_gateway/src/services/broker.rs
use futures::StreamExt;
use lapin::{
    options::{
        BasicAckOptions, BasicConsumeOptions, BasicPublishOptions, ExchangeDeclareOptions,
        QueueBindOptions, QueueDeclareOptions,
    },
    types::{AMQPValue, FieldTable, LongString},
    BasicProperties, Channel, Connection, ConnectionProperties, ExchangeKind,
};
use serde::Deserialize;
use serde_json::value::RawValue;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::Mutex;
use tracing::warn;

use crate::services::app::AppState;
use crate::services::metrics::Metrics;
use crate::services::message::InternalMessage;
use crate::services::settings::Config;
use crate::services::utils::normalize_json;

#[derive(Clone)]
struct RabbitConfig {
    dsn: String,
    inbox_exchange: String,
    inbox_routing_key: String,
    inbox_queue: String,
    inbox_queue_ttl_ms: i64,
    events_exchange: String,
    events_routing_key: String,
    events_queue: String,
    events_queue_ttl_ms: i64,
    dlq_exchange: String,
    dlq_queue: String,
}

struct RabbitPublishState {
    connection: Option<Connection>,
    channel: Option<Channel>,
}

#[derive(Deserialize)]
struct OutboxMessageRaw {
    subjects: Vec<String>,
    #[serde(default)]
    payload: Option<Box<RawValue>>,
}

fn build_event_text(payload: Option<&RawValue>, strip_internal: bool) -> String {
    let raw = payload.map(|v| v.get()).unwrap_or("null");
    if strip_internal {
        if let Ok(internal) = serde_json::from_str::<InternalMessage>(raw) {
            if let Ok(client_json) = serde_json::to_string(&internal.to_client_payload()) {
                let mut text = String::with_capacity(client_json.len() + 28);
                text.push_str(r#"{"type":"event","payload":"#);
                text.push_str(&client_json);
                text.push('}');
                return text;
            }
        }
    }
    let mut text = String::with_capacity(raw.len() + 28);
    text.push_str(r#"{"type":"event","payload":"#);
    text.push_str(raw);
    text.push('}');
    text
}

pub(crate) struct RabbitPublisher {
    config: RabbitConfig,
    state: Mutex<RabbitPublishState>,
}

impl RabbitPublisher {
    pub(crate) fn new(config: &Config) -> Self {
        Self {
            config: RabbitConfig {
                dsn: config.rabbitmq_dsn.clone(),
                inbox_exchange: config.rabbitmq_inbox_exchange.clone(),
                inbox_routing_key: config.rabbitmq_inbox_routing_key.clone(),
                inbox_queue: config.rabbitmq_inbox_queue.clone(),
                inbox_queue_ttl_ms: config.rabbitmq_inbox_queue_ttl_ms,
                events_exchange: config.rabbitmq_events_exchange.clone(),
                events_routing_key: config.rabbitmq_events_routing_key.clone(),
                events_queue: config.rabbitmq_events_queue.clone(),
                events_queue_ttl_ms: config.rabbitmq_events_queue_ttl_ms,
                dlq_exchange: config.rabbitmq_dlq_exchange.clone(),
                dlq_queue: config.rabbitmq_dlq_queue.clone(),
            },
            state: Mutex::new(RabbitPublishState {
                connection: None,
                channel: None,
            }),
        }
    }

    async fn ensure_channel(&self) -> Option<Channel> {
        {
            let state = self.state.lock().await;
            if let Some(channel) = &state.channel {
                return Some(channel.clone());
            }
        }
        let connection = match Connection::connect(
            &self.config.dsn,
            ConnectionProperties::default(),
        )
        .await
        {
            Ok(conn) => conn,
            Err(err) => {
                warn!("rabbitmq.connect_failed: {err}");
                return None;
            }
        };
        let channel = match connection.create_channel().await {
            Ok(channel) => channel,
            Err(err) => {
                warn!("rabbitmq.channel_failed: {err}");
                return None;
            }
        };
        if let Err(err) = self.declare_topology(&channel).await {
            warn!("rabbitmq.declare_failed: {err}");
        }
        let mut state = self.state.lock().await;
        state.connection = Some(connection);
        state.channel = Some(channel.clone());
        Some(channel)
    }

    async fn declare_topology(&self, channel: &Channel) -> Result<(), lapin::Error> {
        if !self.config.inbox_exchange.is_empty() {
            channel
                .exchange_declare(
                    &self.config.inbox_exchange,
                    ExchangeKind::Direct,
                    ExchangeDeclareOptions { durable: true, ..Default::default() },
                    FieldTable::default(),
                )
                .await?;
        }
        if !self.config.events_exchange.is_empty() {
            channel
                .exchange_declare(
                    &self.config.events_exchange,
                    ExchangeKind::Direct,
                    ExchangeDeclareOptions { durable: true, ..Default::default() },
                    FieldTable::default(),
                )
                .await?;
        }
        if !self.config.dlq_exchange.is_empty() {
            channel
                .exchange_declare(
                    &self.config.dlq_exchange,
                    ExchangeKind::Direct,
                    ExchangeDeclareOptions { durable: true, ..Default::default() },
                    FieldTable::default(),
                )
                .await?;
            if !self.config.dlq_queue.is_empty() {
                channel
                    .queue_declare(
                        &self.config.dlq_queue,
                        QueueDeclareOptions { durable: true, ..Default::default() },
                        FieldTable::default(),
                    )
                    .await?;
                channel
                    .queue_bind(
                        &self.config.dlq_queue,
                        &self.config.dlq_exchange,
                        &self.config.dlq_queue,
                        QueueBindOptions::default(),
                        FieldTable::default(),
                    )
                    .await?;
            }
        }
        if !self.config.inbox_queue.is_empty() {
            let args = rabbit_queue_args(
                self.config.inbox_queue_ttl_ms,
                &self.config.dlq_exchange,
                &self.config.dlq_queue,
            );
            channel
                .queue_declare(
                    &self.config.inbox_queue,
                    QueueDeclareOptions { durable: true, ..Default::default() },
                    args,
                )
                .await?;
            channel
                .queue_bind(
                    &self.config.inbox_queue,
                    &self.config.inbox_exchange,
                    &self.config.inbox_routing_key,
                    QueueBindOptions::default(),
                    FieldTable::default(),
                )
                .await?;
        }
        if !self.config.events_queue.is_empty() {
            let args = rabbit_queue_args(
                self.config.events_queue_ttl_ms,
                &self.config.dlq_exchange,
                &self.config.dlq_queue,
            );
            channel
                .queue_declare(
                    &self.config.events_queue,
                    QueueDeclareOptions { durable: true, ..Default::default() },
                    args,
                )
                .await?;
            channel
                .queue_bind(
                    &self.config.events_queue,
                    &self.config.events_exchange,
                    &self.config.events_routing_key,
                    QueueBindOptions::default(),
                    FieldTable::default(),
                )
                .await?;
        }
        Ok(())
    }

    pub(crate) async fn publish(&self, exchange: &str, routing_key: &str, payload: &Value) -> bool {
        let body = normalize_json(payload.clone());
        let body = serde_json::to_string(&body).unwrap_or_else(|_| "{}".to_string());
        self.publish_raw(exchange, routing_key, body.as_bytes()).await
    }

    pub(crate) async fn publish_raw(&self, exchange: &str, routing_key: &str, body: &[u8]) -> bool {
        if exchange.is_empty() {
            return false;
        }
        let Some(channel) = self.ensure_channel().await else {
            return false;
        };
        let confirm = channel
            .basic_publish(
                exchange,
                routing_key,
                BasicPublishOptions::default(),
                body,
                BasicProperties::default(),
            )
            .await;
        match confirm {
            Ok(confirm) => confirm.await.is_ok(),
            Err(_) => false,
        }
    }
}

fn rabbit_queue_args(ttl_ms: i64, dlq_exchange: &str, dlq_queue: &str) -> FieldTable {
    let mut table = FieldTable::default();
    if ttl_ms > 0 {
        let ttl = ttl_ms.min(u32::MAX as i64) as u32;
        table.insert("x-message-ttl".into(), AMQPValue::LongUInt(ttl));
    }
    if !dlq_exchange.is_empty() {
        table.insert(
            "x-dead-letter-exchange".into(),
            AMQPValue::LongString(LongString::from(dlq_exchange.to_string())),
        );
        if !dlq_queue.is_empty() {
            table.insert(
                "x-dead-letter-routing-key".into(),
                AMQPValue::LongString(LongString::from(dlq_queue.to_string())),
            );
        }
    }
    table
}

pub(crate) async fn redis_outbox_consumer(state: Arc<AppState>) {
    let Some(client) = &state.redis else { return; };
    let mut last_id = if state.config.replay_strategy == "none" {
        "$".to_string()
    } else {
        "0-0".to_string()
    };
    let mut backoff = 1.0;
    loop {
        match client.get_multiplexed_async_connection().await {
            Ok(mut conn) => {
                backoff = 1.0;
                loop {
                    let response: redis::RedisResult<Vec<(String, Vec<(String, HashMap<String, String>)>)>> =
                        redis::cmd("XREAD")
                            .arg("BLOCK")
                            .arg(5000)
                            .arg("COUNT")
                            .arg(10)
                            .arg("STREAMS")
                            .arg(&state.config.redis_stream)
                            .arg(&last_id)
                            .query_async(&mut conn)
                            .await;
                    match response {
                        Ok(streams) => {
                            if streams.is_empty() {
                                continue;
                            }
                            for (_stream, entries) in streams {
                                for (id, fields) in entries {
                                    last_id = id;
                                    if let Some(raw) = fields.get("data") {
                                        if let Ok(value) =
                                            serde_json::from_str::<OutboxMessageRaw>(raw)
                                        {
                                            let text = build_event_text(
                                                value.payload.as_deref(),
                                                state.config.outbox_strip_internal,
                                            );
                                            let stats = state
                                                .connections
                                                .send_text_to_subjects(&value.subjects, &text)
                                                .await;
                                            if stats.sent > 0 {
                                                Metrics::inc(
                                                    &state.metrics.ws_messages_out_total,
                                                    stats.sent as u64,
                                                );
                                            }
                                            if stats.dropped > 0 {
                                                Metrics::inc(
                                                    &state.metrics.backpressure_dropped_total,
                                                    stats.dropped as u64,
                                                );
                                            }
                                        }
                                    }
                                }
                            }
                        }
                        Err(err) => {
                            warn!("redis.outbox.error: {err}");
                            break;
                        }
                    }
                }
            }
            Err(err) => {
                warn!("redis.connect_failed: {err}");
            }
        }
        tokio::time::sleep(Duration::from_secs_f64(backoff)).await;
        backoff = (backoff * 2.0).min(30.0);
    }
}

pub(crate) async fn rabbit_outbox_consumer(state: Arc<AppState>) {
    if state.config.rabbitmq_dsn.is_empty() || state.config.rabbitmq_queue.is_empty() {
        return;
    }
    let mut backoff = 1.0;
    loop {
        match Connection::connect(&state.config.rabbitmq_dsn, ConnectionProperties::default()).await
        {
            Ok(connection) => {
                backoff = 1.0;
                let channel = match connection.create_channel().await {
                    Ok(channel) => channel,
                    Err(err) => {
                        warn!("rabbitmq.channel_failed: {err}");
                        continue;
                    }
                };
                if let Err(err) = channel
                    .exchange_declare(
                        &state.config.rabbitmq_exchange,
                        ExchangeKind::Direct,
                        ExchangeDeclareOptions { durable: true, ..Default::default() },
                        FieldTable::default(),
                    )
                    .await
                {
                    warn!("rabbitmq.exchange_failed: {err}");
                    continue;
                }
                if !state.config.rabbitmq_dlq_exchange.is_empty() {
                    let _ = channel
                        .exchange_declare(
                            &state.config.rabbitmq_dlq_exchange,
                            ExchangeKind::Direct,
                            ExchangeDeclareOptions { durable: true, ..Default::default() },
                            FieldTable::default(),
                        )
                        .await;
                    if !state.config.rabbitmq_dlq_queue.is_empty() {
                        let _ = channel
                            .queue_declare(
                                &state.config.rabbitmq_dlq_queue,
                                QueueDeclareOptions { durable: true, ..Default::default() },
                                FieldTable::default(),
                            )
                            .await;
                        let _ = channel
                            .queue_bind(
                                &state.config.rabbitmq_dlq_queue,
                                &state.config.rabbitmq_dlq_exchange,
                                &state.config.rabbitmq_dlq_queue,
                                QueueBindOptions::default(),
                                FieldTable::default(),
                            )
                            .await;
                    }
                }
                let args = rabbit_queue_args(
                    state.config.rabbitmq_queue_ttl_ms,
                    &state.config.rabbitmq_dlq_exchange,
                    &state.config.rabbitmq_dlq_queue,
                );
                if let Err(err) = channel
                    .queue_declare(
                        &state.config.rabbitmq_queue,
                        QueueDeclareOptions { durable: true, ..Default::default() },
                        args,
                    )
                    .await
                {
                    warn!("rabbitmq.queue_failed: {err}");
                    continue;
                }
                if let Err(err) = channel
                    .queue_bind(
                        &state.config.rabbitmq_queue,
                        &state.config.rabbitmq_exchange,
                        &state.config.rabbitmq_routing_key,
                        QueueBindOptions::default(),
                        FieldTable::default(),
                    )
                    .await
                {
                    warn!("rabbitmq.bind_failed: {err}");
                    continue;
                }

                let consumer = match channel
                    .basic_consume(
                        &state.config.rabbitmq_queue,
                        "rust_gateway_outbox",
                        BasicConsumeOptions::default(),
                        FieldTable::default(),
                    )
                    .await
                {
                    Ok(consumer) => consumer,
                    Err(err) => {
                        warn!("rabbitmq.consume_failed: {err}");
                        continue;
                    }
                };

                let mut consumer = consumer;
                while let Some(delivery) = consumer.next().await {
                    match delivery {
                        Ok(delivery) => {
                            let raw = delivery.data.clone();
                            if let Ok(value) = serde_json::from_slice::<OutboxMessageRaw>(&raw) {
                                let text = build_event_text(
                                    value.payload.as_deref(),
                                    state.config.outbox_strip_internal,
                                );
                                let stats = state
                                    .connections
                                    .send_text_to_subjects(&value.subjects, &text)
                                    .await;
                                if stats.sent > 0 {
                                    Metrics::inc(
                                        &state.metrics.ws_messages_out_total,
                                        stats.sent as u64,
                                    );
                                }
                                if stats.dropped > 0 {
                                    Metrics::inc(
                                        &state.metrics.backpressure_dropped_total,
                                        stats.dropped as u64,
                                    );
                                }
                            } else {
                                push_rabbit_dlq(&channel, &state.config, "parse_error", &raw).await;
                            }
                            let _ = delivery.ack(BasicAckOptions::default()).await;
                        }
                        Err(err) => {
                            warn!("rabbitmq.delivery_failed: {err}");
                            break;
                        }
                    }
                }
            }
            Err(err) => {
                warn!("rabbitmq.connect_failed: {err}");
            }
        }
        tokio::time::sleep(Duration::from_secs_f64(backoff)).await;
        backoff = (backoff * 2.0).min(30.0);
    }
}

async fn push_rabbit_dlq(channel: &Channel, config: &Config, reason: &str, raw: &[u8]) {
    if config.rabbitmq_dlq_exchange.is_empty() || config.rabbitmq_dlq_queue.is_empty() {
        return;
    }
    let _ = channel
        .exchange_declare(
            &config.rabbitmq_dlq_exchange,
            ExchangeKind::Direct,
            ExchangeDeclareOptions { durable: true, ..Default::default() },
            FieldTable::default(),
        )
        .await;
    let _ = channel
        .queue_declare(
            &config.rabbitmq_dlq_queue,
            QueueDeclareOptions { durable: true, ..Default::default() },
            FieldTable::default(),
        )
        .await;
    let _ = channel
        .queue_bind(
            &config.rabbitmq_dlq_queue,
            &config.rabbitmq_dlq_exchange,
            &config.rabbitmq_dlq_queue,
            QueueBindOptions::default(),
            FieldTable::default(),
        )
        .await;
    let mut headers = FieldTable::default();
    headers.insert(
        "reason".into(),
        AMQPValue::LongString(LongString::from(reason.to_string())),
    );
    let props = BasicProperties::default().with_headers(headers);
    if let Ok(confirm) = channel
        .basic_publish(
            &config.rabbitmq_dlq_exchange,
            &config.rabbitmq_dlq_queue,
            BasicPublishOptions::default(),
            raw,
            props,
        )
        .await
    {
        let _ = confirm.await;
    }
}
